query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+            DatomsTable::FulltextValues => "fulltext_values",
+            DatomsTable::FulltextDatoms => "fulltext_datoms",
+            DatomsTable::AllDatoms => "all_datoms",
+        }
+    }
+}
+
+/// One of the named columns of our tables.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum DatomsColumn {
+    Entity,
+    Attribute,
+    Value,
+    Tx,
+    ValueTypeTag,
+}
+
+/// A specific instance of a table within a query. E.g., "datoms123".
+pub type TableAlias = String;
+
+/// The association between a table and its alias. E.g., AllDatoms, "all_datoms123".
+#[derive(PartialEq, Eq, Debug)]
+pub struct SourceAlias(DatomsTable, TableAlias);
+
+/// A particular column of a particular aliased table. E.g., "datoms123", Attribute.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct QualifiedAlias(TableAlias, DatomsColumn);
+
+impl QualifiedAlias {
+    fn for_type_tag(&self) -> QualifiedAlias {
+        QualifiedAlias(self.0.clone(), DatomsColumn::ValueTypeTag)
+    }
+
+    #[inline]
+    fn is_value(&self) -> bool {
+        self.1 == DatomsColumn::Value
+    }
+}
+
+/// A thing that's capable of aliasing a table name for us.
+/// This exists so that we can obtain predictable names in tests.
+pub type TableAliaser = Box<FnMut(DatomsTable) -> TableAlias>;
+
+pub fn default_table_aliaser() -> TableAliaser {
+    let mut i = -1;
+    Box::new(move |table| {
+        i += 1;
+        format!("{}{}", table.name(), i)
+    })
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ColumnConstraint {
+    EqualsEntity(QualifiedAlias, Entid),
+    EqualsValue(QualifiedAlias, TypedValue),
+    EqualsColumn(QualifiedAlias, QualifiedAlias),
+}
+
+pub struct ConjoiningClauses {
+    is_known_empty: bool,
+    aliaser: TableAliaser,
+    from: Vec<SourceAlias>,
+    wheres: Vec<ColumnConstraint>,
+    bindings: BTreeMap<Variable, Vec<QualifiedAlias>>,
+    known_types: BTreeMap<Variable, ValueType>,
+    extracted_types: BTreeMap<Variable, QualifiedAlias>,
+}
+
+impl ::std::fmt::Debug for ConjoiningClauses {
+    fn fmt(&self, fmt: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        fmt.debug_struct("ConjoiningClauses")
+            .field("is_known_empty", &self.is_known_empty)
+            .field("from", &self.from)
+            .field("wheres", &self.wheres)
+            .field("bindings", &self.bindings)
+            .field("known_types", &self.known_types)
+            .field("extracted_types", &self.extracted_types)
+            .finish()
+    }
+}
+
+/// Basics.
+impl ConjoiningClauses {
+    pub fn new() -> ConjoiningClauses {
+        ConjoiningClauses {
+            is_known_empty: false,
+            aliaser: default_table_aliaser(),
+            from: vec![],
+            wheres: vec![],
+            bindings: BTreeMap::new(),
+            known_types: BTreeMap::new(),
+            extracted_types: BTreeMap::new(),
+        }
+    }
+
+    pub fn bind_column_to_var(&mut self, table: TableAlias, column: DatomsColumn, var: Variable) {
```

Comment:
Yes. The CC pipeline was always "Clojure-in-place" — that is, completely chained and consuming — with the exception of certain split-and-merge things like complex `or-join`. Not as big a jump to Rust as one might think.